url/url_test.cpp

```diff
@@ -634,6 +637,105 @@ int main() {
         etest::expect_eq(*url, url::Url{.scheme = "a", .host = url::Host{.type = url::HostType::Opaque}});
     });
 
+    etest::test("Web Platform Tests", [] {
+        url::UrlParser p;
+
+        simdjson::ondemand::parser parser;
+
+        auto json = simdjson::padded_string::load("external/wpt/url/resources/urltestdata.json");
+
+        simdjson::ondemand::document doc = parser.iterate(json);
+
+        simdjson::ondemand::array arr = doc.get_array();
+
+        for (auto obj : arr) {
+            // Skip strings, those are just comments
+            if (obj.type() == simdjson::ondemand::json_type::string) {
+                continue;
+            }
+
+            bool failure = false;
+
+            // Check if test expects failure
+            if (obj.find_field("failure").error() != simdjson::error_code::NO_SUCH_FIELD) {
+                failure = true;
+            }
+
+            // Get input URL
+            std::string_view input = obj["input"].get_string(true);
+
+            // Parse base URL if it exists
+            std::optional<url::Url> base_test;
+
+            if (!obj["base"].is_null()) {
+                std::string_view base_str = obj["base"].get_string(true);
+
+                base_test = p.parse(std::string{base_str});
+
+                if (!failure) {
+                    etest::expect(base_test.has_value(), "Parsing base URL:(" + std::string{base_str} + ") failed");
+
+                    continue;
+                }
+            }
+
+            // Parse input URL
+            std::optional<url::Url> url = p.parse(std::string{input}, base_test);
+
+            if (!failure) {
+                etest::expect(url.has_value(), "Parsing input URL:(" + std::string{input} + ") failed");
```

Comment:
You need a continue after this, or we hit UB when the rest of the checks in the test dereference `url`.

Reply:
I added another check to gate the continue; is there a nicer way to use the expect() to avoid doing the comparison again? Doesn't look like it. Maybe we can make the etest functions return a boolean or something?